pkg/limits/stream_metadata.go

```diff
@@ -103,6 +107,83 @@ func (s *streamMetadata) Usage(tenant string, fn UsageFunc) {
 		}
 	}
 }
+
+func (s *streamMetadata) TryStore(tenant string, streams map[int32][]Stream, limit uint64, bucketStart, bucketCutOff int64) []Stream {
+	i := s.getStripeIdx(tenant)
+
+	s.locks[i].Lock()
+	defer s.locks[i].Unlock()
+
+	dropped := make([]Stream, 0)
+	for partitionID, streams := range streams {
+		// Get the current number of active streams for this tenant and partition
+		activeStreams := len(s.stripes[i][tenant][partitionID])
+
+		// Get the maximum number of active streams that can be stored
+		remaining := int(limit) - activeStreams
+
+		// Drop streams that exceed the limit
+		dropped = append(dropped, streams[remaining:]...)
+
+		// Store the streams that are within the limit
+		for _, stream := range streams[:remaining] {
+			// Check if the stream already exists in the metadata
+			if recorded, ok := s.stripes[i][tenant][partitionID][stream.Hash]; ok {
+				// Update total size
+				totalSize := stream.TotalSize + recorded.TotalSize
+
+				// Update or add size for the current bucket
+				updated := false
+				sb := make([]RateBucket, 0, len(stream.RateBuckets)+1)
+
+				// Only keep buckets within the rate window and update the current bucket
```

Comment:
Isn't all this logic duplicated with `Store`? We should re-use it between both `Store` and `TryStore` 👍